### ISSUE TITLE
Add groq API

### ISSUE THREAD

**Issue by iamnvt** [cursor/cursor#1407]
this is the game changer for the speed with llama 3.

**Comment by spikecodes** [comment]
+1 for this ^

**Comment by yangcheng** [comment]
it can be as simple as let cursor respect 'OPENAI_API_BASE_URL' env

**Comment by iamnvt** [comment]
But it will make you lose the access to openai, and also dont have codebase coding.
________________________________

**Comment by CalamariDude** [comment]
This would be awesome - has anyone figured out a workaround for now it would be a gamechanger prob 10x my dev speed

**Comment by yangcheng** [comment]
> This would be awesome - has anyone figured out a workaround for now it would be a gamechanger prob 10x my dev speed

yes! check my twitter, https://twitter.com/yangcheng/status/1782966529057050797  , you need to change a sqlite file but I think developers know what they are doing

**Comment by CalamariDude** [comment]
@yangcheng Thank you! For some reason the speed seems around the same as current gpt4 in cursor, is this your experience?

**Comment by yangcheng** [comment]
for me the chat is noticeably faster. commnad+K is about same, maybe the bottleneck is somewhere else. maybe you can try 8b models to be sure?

**Comment by CalamariDude** [comment]
> for me the chat is noticeably faster. commnad+K is about same, maybe the bottleneck is somewhere else. maybe you can try 8b models to be sure?

its about the same for small models on groq : /

**Comment by kcolemangt** [comment]
Today I made [llm-router](https://x.com/KColemanGT/status/1784798270574801399), enabling the use of `⌘ K` or `⌘ L`, followed by `⌘ /`, to toggle between openai and groq. If there's interest, I can open source it and cut a release.

**Comment by spikecodes** [comment]
@yangcheng Thanks! But how do we pass our API key? https://console.groq.com/docs/api-keys says:

> API keys are required for accessing the APIs.

**Comment by kcolemangt** [comment]
> @yangcheng Thanks! But how do we pass our API key? https://console.groq.com/docs/api-keys says:
> 
> > API keys are required for accessing the APIs.

Set the base URL and API key to Groq. However, you will lose access to OpenAI models until you remove the base URL and replace your Groq key with your OpenAI key.

The llm-router option I mentioned [above](https://github.com/getcursor/cursor/issues/1407#issuecomment-2081879261) allows you to provide the Groq API key via an environment variable, enabling you to use OpenAI models and others, like OLLAMA, simultaneously.

**Comment by spikecodes** [comment]
Thanks @kcolemangt! I didn't realize, I could pass the Groq API key the same as the OpenAI API key.

Regarding llm-router, I don't think it would help me much as I don't really need to switch back and forth between models. Just want a good fast option.

**Comment by CalamariDude** [comment]
Does anyone know why cursor is still slow even using groq? I would imagine the api is slightly faster if the gpt part of the response is done in 1/8th of the time... is this a limitation of extensions built with vscodium ? increasing the speed of cursor seems like the next important milestone to get this to become the best AI editor. @truell20 @Sanger2000 Do you know who would be the best person to ask regarding this ?

**Comment by yangcheng** [comment]
> > @yangcheng Thanks! But how do we pass our API key? https://console.groq.com/docs/api-keys says:
> > > API keys are required for accessing the APIs.
> 
> @spikecodes Set the base URL and API key to Groq. However, you will lose access to OpenAI models until you remove the base URL and replace your Groq key with your OpenAI key.
> 
> The llm-router option I mentioned [above](https://github.com/getcursor/cursor/issues/1407#issuecomment-2081879261) allows you to provide the Groq API key via an environment variable, enabling you to use OpenAI models and others, like OLLAMA, simultaneously.

saw you tweet, very cool,how can we use it?

**Comment by krishnapraveen7** [comment]
> Today I made [llm-router](https://x.com/KColemanGT/status/1784798270574801399), enabling the use of `⌘ K` or `⌘ L`, followed by `⌘ /`, to toggle between openai and groq. If there's interest, I can open source it and cut a release.

Yes please, god i need this

**Comment by kcolemangt** [comment]
https://github.com/kcolemangt/llm-router

**Comment by aicuby** [comment]
First time using cursor - implemented open source groq. How does the automation work with cursor setting up folders, files, python packages, and so on automatically without the need to manually do it yourself?

**Comment by sugoidesune** [comment]
Considering the API is compatible with OpenAI's all we need is another tab to not have to manually change urls and api keys.
![{45D83373-F982-4CD4-BE53-BF85A8300F76}](https://github.com/user-attachments/assets/c6db84d0-3311-400a-a54c-9bac649a8d8d)

**Comment by marko911** [comment]
I've added Groq API and the key but it fails, it says invalid api key but it works fine when I curl. Anyone have same issue?

**Comment by psmyrdek** [comment]
Seems like recently introduced changes for override URL. I can fetch openrouter but nothing else (had local solution based on https://github.com/przeprogramowani/cursor-ollama-backend) and it stopped working.

**Comment by hakangit** [comment]
It fails now because it tries to verify using an openai model like gpt-4o-mini and groq doesn't have those. that makes the verification fail. it would need to be hacked around now if this is to work.

**Comment by Eforen** [comment]
I am also getting the same failure as @hakangit a way to map models to dif providers would be amazing particularly if we can have a display name that is diff from the actual model name.

**Comment by kcolemangt** [comment]
@hakangit @Eforen I use [llm-router](https://github.com/kcolemangt/llm-router/releases) to add Groq and Ollama providers. v0.1.0 adds support for Cursor’s verification checks.

**Comment by Eforen** [comment]
Fair point

**Comment by Viranchee** [comment]
The trick is:
1. Disable all existing models (& remove if possible)
2. Add Groq Models [from here](https://console.groq.com/docs/rate-limits)
3. Add Groq API key [from here](https://console.groq.com/keys)
4. Add Groq URL
5. Hit connect

<img width="763" alt="Image" src="https://github.com/user-attachments/assets/0b2ad4d4-b06e-4dcf-a5a8-f6c982bc8b27" />
